File: java/socketd/src/main/java/org/noear/socketd/transport/core/entity/MessageDefault.java
package org.noear.socketd.transport.core.entity;

import org.noear.socketd.transport.core.*;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Map;


/**
 * 消息默认实现（帧[消息[实体]]）
 *
 * @author noear
 * @since 2.0
 */
public class MessageDefault implements MessageInternal {
    private final int flag;
    private final String sid;
    private final String event;
    private final Entity entity;

    public MessageDefault(int flag, String sid, String event, Entity entity) {
        this.flag = flag;
        this.sid = sid;
        this.event = event;
        this.entity = entity;
    }

    /**
     * 获取标记
     */
    @Override
    public int flag() {
        return flag;
    }

    /**
     * 是否答复结束
     */
    @Override
    public boolean isEnd() {
        return flag == Flags.ReplyEnd;
    }

    /**
     * 是否为请求
     */
    @Override
    public boolean isRequest() {
        return flag == Flags.Request;
    }

    /**
     * 是否为订阅
     */
    @Override
    public boolean isSubscribe() {
        return flag == Flags.Subscribe;
    }

    /**
     * 获取消息流Id（用于消息交互、分片）
     */
    @Override
    public String sid() {
        return sid;
    }

    /**
     * 获取消息事件
     */
    @Override
    public String event() {
        return event;
    }

    /**
     * 获取消息实体
     */
    @Override
    public Entity entity() {
        return entity;
    }

    @Override
    public String toString() {
        return "Message{" +
                "sid='" + sid + '\'' +
                ", event='" + event + '\'' +
                ", entity=" + entity +
                '}';
    }

    @Override
    public String metaString() {
        return entity.metaString();
    }

    @Override
    public Map<String, String> metaMap() {
        return entity.metaMap();
    }

    @Override
    public String meta(String name) {
        return entity.meta(name);
    }

    @Override
    public String metaOrDefault(String name, String def) {
        return entity.metaOrDefault(name, def);
    }

    @Override
    public void putMeta(String name, String val) {
        entity.putMeta(name, val);
    }

    @Override
    public void delMeta(String name) {
        entity.delMeta(name);
    }

    @Override
    public ByteBuffer data() {
        return entity.data();
    }

    @Override
    public String dataAsString() {
        return entity.dataAsString();
    }

    @Override
    public byte[] dataAsBytes() {
        return entity.dataAsBytes();
    }

    @Override
    public int dataSize() {
        return entity.dataSize();
    }

    @Override
    public void release() throws IOException {
        if (entity != null) {
            entity.release();
        }
    }
}

File: python/socketd/socketd/transport/core/entity/MessageDefault.py
from io import BytesIO
from typing import Optional, Any, Dict

from socketd.transport.core.Entity import Entity
from socketd.transport.core.Flags import Flags
from socketd.transport.core.Message import MessageInternal


class MessageDefault(MessageInternal):
    def __init__(self, flag:int, sid:str, event:str, entity:Entity):
        self._flag = flag
        self._sid = sid
        self._event = event
        self._entity = entity

    def flag(self):
        return self._flag

    def is_end(self):
        return self._flag == Flags.ReplyEnd

    def is_request(self):
        return self._flag == Flags.Request

    def is_subscribe(self):
        return self._flag == Flags.Subscribe

    def is_close(self):
        return self._flag == Flags.Close

    def sid(self):
        return self._sid

    def event(self):
        return self._event

    def entity(self):
        return self._entity

    def __str__(self):
        return f"Message{{sid='{self._sid}', event='{self._event}', entity={self._entity}}}"

    def meta_string(self) -> str:
        return self._entity.meta_string()

    def meta_map(self) -> Dict[str, str]:
        return self._entity.meta_map()

    def meta(self, name: str) -> Optional[Any]:
        return self._entity.meta(name)

    def meta_or_default(self, name: str, default: str) -> str:
        return self._entity.meta_or_default(name, default)

    def meta_as_int(self, name:str) ->int:
        return self._entity.meta_as_int(name)

    def put_meta(self, name:str, val:str):
        return self._entity.put_meta(name, val)

    def del_meta(self, name:str):
        return self._entity.del_meta(name)

    def data(self) -> BytesIO:
        return self._entity.data()

    def data_as_string(self) -> str:
        return self._entity.data_as_string()


    def data_as_bytes(self) -> bytes:
        return self._entity.data_as_bytes()

    def data_size(self) -> int:
        return self._entity.data_size()

    def release(self):
        if self._entity:
            self._entity.release()

    def __str__(self):
        return f"Message(sid='{self._sid}', event='{self._event}', entity='{self._entity}')"


